### PR TITLE
Go: Pool connections using Peer/SubChannel abstractions

### PR DESCRIPTION
**Pull request by prashantv**
Basic connection pooling/subchannel support:
 - pool connections and reuse them if possible
 - move mutable channel state into a separate mutable struct (to make it easy to avoid accidentally accessing it without the lock)

The OnActive event from connection is kind of hacky, if we need more event support in future, we can create a proper way to register for events.

### CODE REVIEW

File: golang/peer.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+	"golang.org/x/net/context"
+)
+
+var (
+	errInvalidConnectionState = errors.New("connection is in an invalid state")
+)
+
+// PeerList maintains a list of Peers.
+type PeerList struct {
```
Comment (mmihic) [inline golang/peer.go]: Does this need to be exported?

**Comment by mmihic** [comment]
Nice.  Maybe consider having a test that does concurrent peer / connection setup, to run under the race detector.

I think the OnActive thing is fine, there's precedence for taking this type of simple approach in http.Server#ConnState.

File: golang/peer.go (hunk 1 above)
Comment (prashantv) [inline golang/peer.go]: Yep, exported it, thanks.

**Comment by prashantv** [comment]
Yep, I'm planning to write a stress test with lots of concurrent connections/calls which we can run with the race detector. However, I want to consolidate some of the test setup logic that could be shared as part of that, so I'll do that separately.

**Pull request merged**